core/server/master/src/main/java/alluxio/master/metastore/CachingInodeStore.java

```diff
@@ -0,0 +1,143 @@
+import com.google.common.cache.CacheBuilder;
+
+import java.util.Iterator;
+import java.util.Optional;
+
+/**
+ * An inode store which caches inode tree metadata and delegates to another inode store for cache
+ * misses.
+ */
+public final class CachingInodeStore implements InodeStore {
+  private final InodeStore mBackingStore;
+
+  // Cache recently-accessed inodes.
+  private final Cache<Long, MutableInode<?>> mInodeCache = CacheBuilder.newBuilder()
+      .maximumSize(Configuration.getInt(PropertyKey.MASTER_METASTORE_INODE_CACHE_SIZE)).build();
+
+  // Cache recently-accessed inode tree edges.
+  private final Cache<Edge, Long> mEdgeCache = CacheBuilder.newBuilder()
+      .maximumSize(Configuration.getInt(PropertyKey.MASTER_METASTORE_INODE_CACHE_SIZE)).build();
+
+  /**
+   * @param backingStore the backing inode store
+   */
+  public CachingInodeStore(InodeStore backingStore) {
+    mBackingStore = backingStore;
+  }
+
+  @Override
+  public Optional<MutableInode<?>> getMutable(long id) {
+    Optional<MutableInode<?>> inode = Optional.ofNullable(mInodeCache.getIfPresent(id));
+    if (inode.isPresent()) {
+      return inode;
+    }
+    inode = mBackingStore.getMutable(id);
+    if (inode.isPresent()) {
+      mInodeCache.put(id, inode.get());
+    }
+    return inode;
+  }
+
+  @Override
+  public void remove(InodeView inode) {
+    mInodeCache.invalidate(inode.getId());
+    mEdgeCache.invalidate(new Edge(inode.getParentId(), inode.getName()));
+    mBackingStore.remove(inode);
+  }
+
+  @Override
+  public void writeInode(MutableInode<?> inode) {
+    mInodeCache.put(inode.getId(), inode);
+    mBackingStore.writeInode(inode);
+  }
+
+  @Override
+  public void clear() {
+    mInodeCache.invalidateAll();
+    mEdgeCache.invalidateAll();
+    mBackingStore.clear();
+  }
+
+  @Override
+  public void addChild(long parentId, InodeView inode) {
+    mEdgeCache.put(new Edge(parentId, inode.getName()), inode.getId());
+    mBackingStore.addChild(parentId, inode);
+  }
+
+  @Override
+  public void removeChild(long parentId, String name) {
+    mEdgeCache.invalidate(new Edge(parentId, name));
+    mBackingStore.removeChild(parentId, name);
+  }
+
+  @Override
+  public long estimateSize() {
+    return mBackingStore.estimateSize();
+  }
+
+  @Override
+  public Iterable<Long> getChildIds(InodeDirectoryView inode) {
+    return mBackingStore.getChildIds(inode);
+  }
+
+  @Override
+  public Iterable<? extends Inode> getChildren(InodeDirectoryView inode) {
+    Iterator<Long> baseIterator = mBackingStore.getChildIds(inode).iterator();
+    return () -> new Iterator<Inode>() {
+      @Override
+      public boolean hasNext() {
+        return baseIterator.hasNext();
+      }
+
+      @Override
+      public Inode next() {
+        return get(baseIterator.next()).get();
+      }
+    };
+  }
+
+  @Override
+  public Optional<Long> getChildId(InodeDirectoryView inode, String name) {
+    return mBackingStore.getChildId(inode, name);
```

Comment:
We can't try for a cache hit here?

Reply:
added a cache hit attempt. It won't affect performance though because this method is never used outside of `CachingInodeStore`